src/Elastic.Apm/Report/PayloadSenderV2.cs

```diff
@@ -97,56 +118,130 @@ internal class PayloadSenderV2 : IPayloadSender, IDisposable
+			if (newEventQueueCount > _maxQueueEventCount)
+			{
+				_logger.Debug()
+					?.Log("Queue reached max capacity - " + dbgEventKind + " will be discarded. "
+						+ " newEventQueueCount: {EventQueueCount}."
+						+ " " + dbgEventKind + ": {" + dbgEventKind + "}."
+						, newEventQueueCount, eventObj);
+				Interlocked.Decrement(ref _eventQueueCount);
+				return false;
+			}
+
+			var enqueuedSuccessfully = _eventQueue.Post(eventObj);
+			if (!enqueuedSuccessfully)
+			{
+				_logger.Debug()
+					?.Log("Failed to enqueue " + dbgEventKind + "."
+						+ " newEventQueueCount: {EventQueueCount}."
+						+ " " + dbgEventKind + ": {" + dbgEventKind + "}."
+						, newEventQueueCount, eventObj);
+				Interlocked.Decrement(ref _eventQueueCount);
+				return false;
+			}
+
 			_logger.Debug()
-				?.Log(!res
-					? "Failed adding Error to the queue, {Error}"
-					: "Error added to the queue, {Error}", error);
+				?.Log("Enqueued " + dbgEventKind + "."
+					+ " newEventQueueCount: {EventQueueCount}."
+					+ " " + dbgEventKind + ": {" + dbgEventKind + "}."
+					, newEventQueueCount, eventObj);
+
+			if (_flushInterval == TimeSpan.Zero) _eventQueue.TriggerBatch();
+
+			return true;
 		}
 
 		private async Task DoWork()
 		{
+			var minTimeBetweenWaitingOnTimerLogs = TimeSpan.FromSeconds(10);
+			var stopwatch = Stopwatch.StartNew();
+			TimeSpan? elapsedOnLastWaitingOnTimerLog = null;
+
 			_batchBlockReceiveAsyncCts = new CancellationTokenSource();
+			Task<object[]> receiveAsyncTask = null;
 			while (true)
 			{
-				var queueItems = await _eventQueue.ReceiveAsync(_batchBlockReceiveAsyncCts.Token);
-				await ProcessQueueItems(queueItems);
+				// ReSharper disable once InconsistentlySynchronizedField
+				if (receiveAsyncTask == null) receiveAsyncTask = _eventQueue.ReceiveAsync(_batchBlockReceiveAsyncCts.Token);
+
+				object[] eventBatchToSend = null;
+				if (_flushInterval == TimeSpan.Zero)
+				{
+					_logger.Trace()?.Log("Waiting for data to send... (not using FlushInterval timer because FlushInterval is 0)");
+					eventBatchToSend = await receiveAsyncTask;
+					receiveAsyncTask = null;
+				}
+				else
+				{
+					var logTimerRelatedInfo = false;
+					var elapsedTime = stopwatch.Elapsed;
+					if (!elapsedOnLastWaitingOnTimerLog.HasValue ||
+						elapsedOnLastWaitingOnTimerLog.Value + minTimeBetweenWaitingOnTimerLogs <= elapsedTime)
+					{
+						logTimerRelatedInfo = true;
+						_logger.Trace()
+							?.Log("Waiting for data to send or FlushInterval timer to be triggered (whichever is earlier)..."
+								+ " _flushInterval: {FlushInterval}", _flushInterval);
+						elapsedOnLastWaitingOnTimerLog = elapsedTime;
+					}
+
+					var flushIntervalDelayTask = Task.Delay((int)_flushInterval.TotalMilliseconds, _batchBlockReceiveAsyncCts.Token);
+					var completedTask = await Task.WhenAny(receiveAsyncTask, flushIntervalDelayTask);
+					if (completedTask == receiveAsyncTask)
```

Comment:
> Shouldn't this be `if (receiveAsyncTask.IsCompleted)`? 

This is the idiomatic way to use `WhenAny` API, for example see https://devblogs.microsoft.com/pfxteam/how-do-i-cancel-non-cancelable-async-operations/
But I changed 

> The `completedTask` still can be `Cancelled` or `Faulted` and that means 2 lines later the `await receiveAsyncTask` will probably also not be successful.

I assume then that in `if (receiveAsyncTask.IsCompleted)` above you actually meant `if (receiveAsyncTask.IsCompletedSuccessfully)` because `IsCompleted` is `true` for `Cancelled` or `Faulted`. I agree that we should be careful about allowing `await receiveAsyncTask` to throw and cause the whole payload sender's thread to stop. I've added `try/catch` to prevent the exception from propagating out the payload sender's thread and `PayloadSenderV2.DoWork`. The latter for the near future when we will share a worker thread pool to run multiple components so we wouldn't want cancellation/exception in one of the components to cause a worker thread to exit. Going back to `await receiveAsyncTask` - AFAICS the only reason for it to throw is cancellation and since we don't need to support cancelling individual receive operations ti follow that the cancellation is for the whole payload sender component. So it's okay to let the cancellation exception propagate out of `await receiveAsyncTask` and cause `PayloadSenderV2.DoWork` to exit (`DoWork` will log the cancellation exception, swallow it and exit normally).

Reply:
ok